Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h

```diff
@@ -112,14 +151,17 @@ void collect_badly_shaped_triangles(const typename boost::graph_traits<TriangleM
 {
   typedef typename boost::graph_traits<TriangleMesh>::halfedge_descriptor       halfedge_descriptor;
 
-  std::array<halfedge_descriptor, 2> res = is_badly_shaped(f, tmesh, vpm, vcm, ecm, gt, cap_threshold,
-                                                           needle_threshold,
+  std::array<halfedge_descriptor, 2> res = is_badly_shaped(f, tmesh, vpm, vcm, ecm, gt,
+                                                           needle_threshold, cap_threshold,
                                                            collapse_length_threshold, flip_triangle_height_threshold_squared);
 
   if(res[0] != boost::graph_traits<TriangleMesh>::null_halfedge())
   {
 #ifdef CGAL_PMP_DEBUG_REMOVE_DEGENERACIES_EXTRA
-    std::cout << "add new needle: " << edge(res[0], tmesh) << std::endl;
+    if (res[1] == boost::graph_traits<TriangleMesh>::null_halfedge())
+      std::cout << "add new needle: " << edge(res[0], tmesh) << std::endl;
+    else
+      std::cout << "add new needle (also a cap): " << edge(res[0], tmesh) << std::endl;
```

Comment:
Do you not fill both ranges on purpose? There's still:

```
  ...
  else // let's not make it possible to have a face be both a cap and a needle (for now)
 ...
```

a little below; so, if it's a needle & a cap, you fill the needle range, but not the cap range.